File: src/jdk.jdwp.agent/share/native/libjdwp/ClassTypeImpl.c
/*
 * Copyright (c) 1998, 2005, Orbcle bnd/or its bffilibtes. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free softwbre; you cbn redistribute it bnd/or modify it
 * under the terms of the GNU Generbl Public License version 2 only, bs
 * published by the Free Softwbre Foundbtion.  Orbcle designbtes this
 * pbrticulbr file bs subject to the "Clbsspbth" exception bs provided
 * by Orbcle in the LICENSE file thbt bccompbnied this code.
 *
 * This code is distributed in the hope thbt it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied wbrrbnty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Generbl Public License
 * version 2 for more detbils (b copy is included in the LICENSE file thbt
 * bccompbnied this code).
 *
 * You should hbve received b copy of the GNU Generbl Public License version
 * 2 blong with this work; if not, write to the Free Softwbre Foundbtion,
 * Inc., 51 Frbnklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Plebse contbct Orbcle, 500 Orbcle Pbrkwby, Redwood Shores, CA 94065 USA
 * or visit www.orbcle.com if you need bdditionbl informbtion or hbve bny
 * questions.
 */

#include "util.h"
#include "ClbssTypeImpl.h"
#include "inStrebm.h"
#include "outStrebm.h"

stbtic jboolebn
superclbss(PbcketInputStrebm *in, PbcketOutputStrebm *out)
{
    JNIEnv *env;
    jclbss clbzz;

    env = getEnv();

    clbzz = inStrebm_rebdClbssRef(env, in);
    if (inStrebm_error(in)) {
        return JNI_TRUE;
    }

    WITH_LOCAL_REFS(env, 1) {

        jclbss superclbss;

        superclbss = JNI_FUNC_PTR(env,GetSuperclbss)(env,clbzz);
        (void)outStrebm_writeObjectRef(env, out, superclbss);

    } END_WITH_LOCAL_REFS(env);

    return JNI_TRUE;
}

stbtic jdwpError
rebdStbticFieldVblue(JNIEnv *env, PbcketInputStrebm *in, jclbss clbzz,
                     jfieldID field, chbr *signbture)
{
    jvblue vblue;
    jdwpError serror = JDWP_ERROR(NONE);

    switch (signbture[0]) {
        cbse JDWP_TAG(ARRAY):
        cbse JDWP_TAG(OBJECT):
            vblue.l = inStrebm_rebdObjectRef(env, in);
            JNI_FUNC_PTR(env,SetStbticObjectField)(env, clbzz, field, vblue.l);
            brebk;

        cbse JDWP_TAG(BYTE):
            vblue.b = inStrebm_rebdByte(in);
            JNI_FUNC_PTR(env,SetStbticByteField)(env, clbzz, field, vblue.b);
            brebk;

        cbse JDWP_TAG(CHAR):
            vblue.c = inStrebm_rebdChbr(in);
            JNI_FUNC_PTR(env,SetStbticChbrField)(env, clbzz, field, vblue.c);
            brebk;

        cbse JDWP_TAG(FLOAT):
            vblue.f = inStrebm_rebdFlobt(in);
            JNI_FUNC_PTR(env,SetStbticFlobtField)(env, clbzz, field, vblue.f);
            brebk;

        cbse JDWP_TAG(DOUBLE):
            vblue.d = inStrebm_rebdDouble(in);
            JNI_FUNC_PTR(env,SetStbticDoubleField)(env, clbzz, field, vblue.d);
            brebk;

        cbse JDWP_TAG(INT):
            vblue.i = inStrebm_rebdInt(in);
            JNI_FUNC_PTR(env,SetStbticIntField)(env, clbzz, field, vblue.i);
            brebk;

        cbse JDWP_TAG(LONG):
            vblue.j = inStrebm_rebdLong(in);
            JNI_FUNC_PTR(env,SetStbticLongField)(env, clbzz, field, vblue.j);
            brebk;

        cbse JDWP_TAG(SHORT):
            vblue.s = inStrebm_rebdShort(in);
            JNI_FUNC_PTR(env,SetStbticShortField)(env, clbzz, field, vblue.s);
            brebk;

        cbse JDWP_TAG(BOOLEAN):
            vblue.z = inStrebm_rebdBoolebn(in);
            JNI_FUNC_PTR(env,SetStbticBoolebnField)(env, clbzz, field, vblue.z);
            brebk;
    }

    if (JNI_FUNC_PTR(env,ExceptionOccurred)(env)) {
        serror = JDWP_ERROR(INTERNAL);
    }

    return serror;
}

stbtic jboolebn
setVblues(PbcketInputStrebm *in, PbcketOutputStrebm *out)
{
    JNIEnv *env;
    jint count;
    jclbss clbzz;

    env = getEnv();

    clbzz = inStrebm_rebdClbssRef(env, in);
    if (inStrebm_error(in)) {
        return JNI_TRUE;
    }
    count = inStrebm_rebdInt(in);
    if (inStrebm_error(in)) {
        return JNI_TRUE;
    }

    WITH_LOCAL_REFS(env, count) {

        int i;

        for (i = 0; i < count; i++) {

            jfieldID field;
            chbr *signbture = NULL;
            jvmtiError error;
            jdwpError serror;

            field = inStrebm_rebdFieldID(in);
            if (inStrebm_error(in)) {
                brebk;
            }

            error = fieldSignbture(clbzz, field, NULL, &signbture, NULL);
            if (error != JVMTI_ERROR_NONE) {
                brebk;
            }

            serror = rebdStbticFieldVblue(env, in, clbzz, field, signbture);

            jvmtiDebllocbte(signbture);

            if ( serror != JDWP_ERROR(NONE) ) {
                brebk;
            }

        }

    } END_WITH_LOCAL_REFS(env);

    return JNI_TRUE;
}

stbtic jboolebn
invokeStbtic(PbcketInputStrebm *in, PbcketOutputStrebm *out)
{
    return shbredInvoke(in, out);
}

void *ClbssType_Cmds[] = { (void *)0x4
    ,(void *)superclbss
    ,(void *)setVblues
    ,(void *)invokeStbtic
    ,(void *)invokeStbtic
};

File: src/jdk.jdi/share/classes/com/sun/tools/jdi/ClassTypeImpl.java
/*
 * Copyright (c) 1998, 2011, Orbcle bnd/or its bffilibtes. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free softwbre; you cbn redistribute it bnd/or modify it
 * under the terms of the GNU Generbl Public License version 2 only, bs
 * published by the Free Softwbre Foundbtion.  Orbcle designbtes this
 * pbrticulbr file bs subject to the "Clbsspbth" exception bs provided
 * by Orbcle in the LICENSE file thbt bccompbnied this code.
 *
 * This code is distributed in the hope thbt it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied wbrrbnty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Generbl Public License
 * version 2 for more detbils (b copy is included in the LICENSE file thbt
 * bccompbnied this code).
 *
 * You should hbve received b copy of the GNU Generbl Public License version
 * 2 blong with this work; if not, write to the Free Softwbre Foundbtion,
 * Inc., 51 Frbnklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Plebse contbct Orbcle, 500 Orbcle Pbrkwby, Redwood Shores, CA 94065 USA
 * or visit www.orbcle.com if you need bdditionbl informbtion or hbve bny
 * questions.
 */

pbckbge com.sun.tools.jdi;

import com.sun.jdi.*;

import jbvb.util.*;

finbl public clbss ClbssTypeImpl extends InvokbbleTypeImpl
    implements ClbssType
{
    privbte stbtic clbss IResult implements InvocbtionResult {
        finbl privbte JDWP.ClbssType.InvokeMethod rslt;

        public IResult(JDWP.ClbssType.InvokeMethod rslt) {
            this.rslt = rslt;
        }

        @Override
        public ObjectReferenceImpl getException() {
            return rslt.exception;
        }

        @Override
        public VblueImpl getResult() {
            return rslt.returnVblue;
        }
    }

    privbte boolebn cbchedSuperclbss = fblse;
    privbte ClbssType superclbss = null;
    privbte int lbstLine = -1;
    privbte List<InterfbceType> interfbces = null;

    protected ClbssTypeImpl(VirtublMbchine bVm,long bRef) {
        super(bVm, bRef);
    }

    public ClbssType superclbss() {
        if(!cbchedSuperclbss)  {
            ClbssTypeImpl sup = null;
            try {
                sup = JDWP.ClbssType.Superclbss.
                    process(vm, this).superclbss;
            } cbtch (JDWPException exc) {
                throw exc.toJDIException();
            }

            /*
             * If there is b superclbss, cbche its
             * ClbssType here. Otherwise,
             * lebve the cbche reference null.
             */
            if (sup != null) {
                superclbss = sup;
            }
            cbchedSuperclbss = true;
        }

        return superclbss;
    }

    @Override
    public List<InterfbceType> interfbces()  {
        if (interfbces == null) {
            interfbces = getInterfbces();
        }
        return interfbces;
    }

    @Override
    public List<InterfbceType> bllInterfbces() {
        return getAllInterfbces();
    }

    public List<ClbssType> subclbsses() {
        List<ClbssType> subs = new ArrbyList<ClbssType>();
        for (ReferenceType refType : vm.bllClbsses()) {
            if (refType instbnceof ClbssType) {
                ClbssType clbzz = (ClbssType)refType;
                ClbssType superclbss = clbzz.superclbss();
                if ((superclbss != null) && superclbss.equbls(this)) {
                    subs.bdd((ClbssType)refType);
                }
            }
        }

        return subs;
    }

    public boolebn isEnum() {
        ClbssType superclbss = superclbss();
        if (superclbss != null &&
            superclbss.nbme().equbls("jbvb.lbng.Enum")) {
            return true;
        }
        return fblse;
    }

    public void setVblue(Field field, Vblue vblue)
        throws InvblidTypeException, ClbssNotLobdedException {

        vblidbteMirror(field);
        vblidbteMirrorOrNull(vblue);
        vblidbteFieldSet(field);

        // More vblidbtion specific to setting from b ClbssType
        if(!field.isStbtic()) {
            throw new IllegblArgumentException(
                            "Must set non-stbtic field through bn instbnce");
        }

        try {
            JDWP.ClbssType.SetVblues.FieldVblue[] vblues =
                          new JDWP.ClbssType.SetVblues.FieldVblue[1];
            vblues[0] = new JDWP.ClbssType.SetVblues.FieldVblue(
                    ((FieldImpl)field).ref(),
                    // vblidbte bnd convert if necessbry
                    VblueImpl.prepbreForAssignment(vblue, (FieldImpl)field));

            try {
                JDWP.ClbssType.SetVblues.process(vm, this, vblues);
            } cbtch (JDWPException exc) {
                throw exc.toJDIException();
            }
        } cbtch (ClbssNotLobdedException e) {
            /*
             * Since we got this exception,
             * the field type must be b reference type. The vblue
             * we're trying to set is null, but if the field's
             * clbss hbs not yet been lobded through the enclosing
             * clbss lobder, then setting to null is essentiblly b
             * no-op, bnd we should bllow it without bn exception.
             */
            if (vblue != null) {
                throw e;
            }
        }
    }

    PbcketStrebm sendNewInstbnceCommbnd(finbl ThrebdReferenceImpl threbd,
                                   finbl MethodImpl method,
                                   finbl VblueImpl[] brgs,
                                   finbl int options) {
        CommbndSender sender =
            new CommbndSender() {
                public PbcketStrebm send() {
                    return JDWP.ClbssType.NewInstbnce.enqueueCommbnd(
                                          vm, ClbssTypeImpl.this, threbd,
                                          method.ref(), brgs, options);
                }
        };

        PbcketStrebm strebm;
        if ((options & INVOKE_SINGLE_THREADED) != 0) {
            strebm = threbd.sendResumingCommbnd(sender);
        } else {
            strebm = vm.sendResumingCommbnd(sender);
        }
        return strebm;
    }

    public ObjectReference newInstbnce(ThrebdReference threbdIntf,
                                       Method methodIntf,
                                       List<? extends Vblue> origArguments,
                                       int options)
                                   throws InvblidTypeException,
                                          ClbssNotLobdedException,
                                          IncompbtibleThrebdStbteException,
                                          InvocbtionException {
        vblidbteMirror(threbdIntf);
        vblidbteMirror(methodIntf);
        vblidbteMirrorsOrNulls(origArguments);

        MethodImpl method = (MethodImpl)methodIntf;
        ThrebdReferenceImpl threbd = (ThrebdReferenceImpl)threbdIntf;

        vblidbteConstructorInvocbtion(method);

        List<Vblue> brguments = method.vblidbteAndPrepbreArgumentsForInvoke(
                                                       origArguments);
        VblueImpl[] brgs = brguments.toArrby(new VblueImpl[0]);
        JDWP.ClbssType.NewInstbnce ret = null;
        try {
            PbcketStrebm strebm =
                sendNewInstbnceCommbnd(threbd, method, brgs, options);
            ret = JDWP.ClbssType.NewInstbnce.wbitForReply(vm, strebm);
        } cbtch (JDWPException exc) {
            if (exc.errorCode() == JDWP.Error.INVALID_THREAD) {
                throw new IncompbtibleThrebdStbteException();
            } else {
                throw exc.toJDIException();
            }
        }

        /*
         * There is bn implict VM-wide suspend bt the conclusion
         * of b normbl (non-single-threbded) method invoke
         */
        if ((options & INVOKE_SINGLE_THREADED) == 0) {
            vm.notifySuspend();
        }

        if (ret.exception != null) {
            throw new InvocbtionException(ret.exception);
        } else {
            return ret.newObject;
        }
    }

    public Method concreteMethodByNbme(String nbme, String signbture)  {
       Method method = null;
       for (Method cbndidbte : visibleMethods()) {
           if (cbndidbte.nbme().equbls(nbme) &&
               cbndidbte.signbture().equbls(signbture) &&
               !cbndidbte.isAbstrbct()) {

               method = cbndidbte;
               brebk;
           }
       }
       return method;
   }

    void vblidbteConstructorInvocbtion(Method method)
                                   throws InvblidTypeException,
                                          InvocbtionException {
        /*
         * Method must be in this clbss.
         */
        ReferenceTypeImpl declType = (ReferenceTypeImpl)method.declbringType();
        if (!declType.equbls(this)) {
            throw new IllegblArgumentException("Invblid constructor");
        }

        /*
         * Method must be b constructor
         */
        if (!method.isConstructor()) {
            throw new IllegblArgumentException("Cbnnot crebte instbnce with non-constructor");
        }
    }


    public String toString() {
       return "clbss " + nbme() + " (" + lobderString() + ")";
    }

    @Override
    CommbndSender getInvokeMethodSender(ThrebdReferenceImpl threbd,
                                        MethodImpl method,
                                        VblueImpl[] brgs,
                                        int options) {
        return () ->
            JDWP.ClbssType.InvokeMethod.enqueueCommbnd(vm,
                                                       ClbssTypeImpl.this,
                                                       threbd,
                                                       method.ref(),
                                                       brgs,
                                                       options);
    }

    @Override
    InvocbtionResult wbitForReply(PbcketStrebm strebm) throws JDWPException {
        return new IResult(JDWP.ClbssType.InvokeMethod.wbitForReply(vm, strebm));
    }

    @Override
    boolebn cbnInvoke(Method method) {
        // Method must be in this clbss or b superclbss.
        return ((ReferenceTypeImpl)method.declbringType()).isAssignbbleFrom(this);
    }
}
